### PR TITLE
Adds broccoli persistent filter

### PR DESCRIPTION
**Pull request by twokul**
Improves performance by introducing `broccoli-persistent-filter` and caching.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -22,4 +27,20 @@ DefeatureifyFilter.prototype.processString = function (str) {
 	return defeatureify(str, this.options);
 };
 
+DefeatureifyFilter.prototype.cacheKeyProcessString = function cacheKey(string, relativePath) {
+	return crypto.createHash('md5').update(string, 'utf8').digest('hex');
+};
```
Comment (sindresorhus) [inline index.js]: Why are you overriding this instead of just using the default?

File: index.js [hunk 2]
```diff
@@ -22,4 +27,20 @@ DefeatureifyFilter.prototype.processString = function (str) {
...
+
+DefeatureifyFilter.prototype.optionsHash  = function() {
+	if (!this._optionsHash) {
+		this._optionsHash = crypto.createHash('md5').update(stringify(this.options), 'utf8').digest('hex');
+	}
+
+	return this._optionsHash;
+};
```
Comment (sindresorhus) [inline index.js]: I don't see `optionsHash` anywhere in the code: https://github.com/stefanpenner/broccoli-persistent-filter/blob/125e78ebffb079f72cdfd6e0bb0a9cf9be96cc12/index.js

**Comment by sindresorhus** [comment]
//@stefanpenner

File: package.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
     "node": ">=0.10.0"
   },
   "scripts": {
-    "test": "broccoli build temp && mocha"
+    "test": "rm -rf temp tmp && broccoli build temp && mocha"
```
Comment (sindresorhus) [inline package.json]: Use [`trash`](https://github.com/sindresorhus/trash) so it works cross-platform.

**Comment by stefanpenner** [comment]
Cool re: trash

File: index.js (hunk 2 above)
Comment (twokul) [inline index.js]: good point. will fix

**Comment by twokul** [comment]
@sindresorhus updated.

**Comment by stefanpenner** [comment]
this is still pending some stuff upstream:

- [ ] windows support
- [ ] thorough tests on broccoli-persistent-filter (i started with broccoli-filter tests first, now must circle back)

File: index.js [hunk 4]
```diff
@@ -1,18 +1,23 @@
 'use strict';
-var Filter = require('broccoli-filter');
+
+var crypto = require('crypto');
+var PersistentFilter = require('broccoli-persistent-filter');
 var defeatureify = require('defeatureify');
+var stringify  = require('json-stable-stringify');
```
Comment (stefanpenner) [inline index.js]: whitespace

**Comment by twokul** [comment]
travis failure seems unrelated to the tests, `npm install` fails (node `0.10`).

File: index.js [hunk 5]
```diff
@@ -22,4 +27,20 @@ DefeatureifyFilter.prototype.processString = function (str) {
...
+	return this.optionsHash() + crypto.createHash('md5').update(string, 'utf8').digest('hex');
+};
+
+DefeatureifyFilter.prototype.optionsHash  = function() {
+	if (!this._optionsHash) {
+		this._optionsHash = crypto.createHash('md5').update(stringify(this.options), 'utf8').digest('hex');
+	}
+
+	return this._optionsHash;
+};
```
Comment (sindresorhus) [inline index.js]: I don't see the point of this method. Hashing the options object shouldn't have any noticeable impact on perf.

Comment (stefanpenner) [inline index.js]: ya it's likely pre-mature optimization.

@twokul can you change this in the broccoli-babel-transpiler as well? @sindresorhus is correct.

File: index.js [hunk 6]
```diff
@@ -22,4 +27,20 @@ DefeatureifyFilter.prototype.processString = function (str) {
 	return defeatureify(str, this.options);
 };
 
+DefeatureifyFilter.prototype.cacheKeyProcessString = function cacheKeyProcessString(string, relativePath) {
+	return this.optionsHash() + crypto.createHash('md5').update(string, 'utf8').digest('hex');
```
Comment (sindresorhus) [inline index.js]: Use https://github.com/sindresorhus/md5-hex

Comment (stefanpenner) [inline index.js]: :+1:

File: index.js [hunk 7]
```diff
@@ -22,4 +26,12 @@ DefeatureifyFilter.prototype.processString = function (str) {
 	return defeatureify(str, this.options);
 };
 
+DefeatureifyFilter.prototype.cacheKeyProcessString = function cacheKeyProcessString(string, relativePath) {
+	return md5Hex(string);
```
Comment (stefanpenner) [inline index.js]: LGTM

**Comment by stefanpenner** [comment]
LGTM

File: readme.md [hunk 8]
```diff
@@ -13,7 +16,6 @@ Defeatureify is used in the [ember-dev](https://github.com/emberjs/ember-dev) pa
 $ npm install --save-dev broccoli-defeatureify
 ```
 
-
```
Comment (kevva) [inline readme.md]: Keep the two blank lines between sections.

File: .travis.yml [hunk 9]
```diff
@@ -4,3 +4,6 @@ node_js:
   - 'iojs'
   - '0.12'
   - '0.10'
+before_install:
+  - "npm config set spin false"
+  - "npm install -g npm@^2"
```
Comment (kevva) [inline .travis.yml]: Why?

File: index.js [hunk 10]
```diff
@@ -22,4 +26,12 @@ DefeatureifyFilter.prototype.processString = function (str) {
 	return defeatureify(str, this.options);
 };
 
+DefeatureifyFilter.prototype.cacheKeyProcessString = function cacheKeyProcessString(string, relativePath) {
```
Comment (kevva) [inline index.js]: Keep the functions anonymous like the rest of them.

File: index.js [hunk 11]
```diff
@@ -22,4 +26,12 @@ DefeatureifyFilter.prototype.processString = function (str) {
...
+	return md5Hex(string);
+};
+
+DefeatureifyFilter.prototype.baseDir = function baseDir() {
```
Comment (kevva) [inline index.js]: Same here, keep it anonymous.

File: package.json (hunk 3 above)
Comment (kevva) [inline package.json]: We already use `rimraf` in the tests. We should remove one of them.

File: .travis.yml (hunk 9 above)
Comment (sindresorhus) [inline .travis.yml]: This is usually to fix npm on node 0.10, but tests passed before this, so I don't think it's strictly needed.

Comment (kevva) [inline .travis.yml]: Indeed, but feels awkward to do this at a module level.

Comment (sindresorhus) [inline .travis.yml]: @kevva Uhm, this is for Travis, not for the user.

Comment (stefanpenner) [inline .travis.yml]: this is totally legit change. Otherwise we can't use @sindresorhus npm namespace

Comment (kevva) [inline .travis.yml]: Haha, yes, lol. Sorry.

File: index.js (hunk 10 above)
Comment (twokul) [inline index.js]: happy to change but this produce confusing stack trace if we need to debug

**Comment by sindresorhus** [comment]
Any movement on this?

**Comment by stefanpenner** [comment]
Been busy. I'll likely have time later this week again

**Comment by sindresorhus** [comment]
ping :)

**Comment by twokul** [comment]
@sindresorhus @stefanpenner I started adding tests to `persistent-filter`, will make sure to wrap that PR up by the end of the week

**Comment by sindresorhus** [comment]
Closing for now. Just ping me when this is ready to be reopened ;)

**Comment by stefanpenner** [comment]
We are merging persistent filter into the main filter as an opt in. At which we can revisit this

**Pull request closed without merging**